### PR TITLE
Master timesheet jem

### PR DESCRIPTION
**Pull request by jem-odoo**
Description of the issue/feature this PR addresses:
https://www.odoo.com/web#id=34702&view_type=form&model=project.task&action=333&active_id=965&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_timesheet/controllers/main.py [hunk 1]
```diff
@@ -12,74 +12,87 @@ class SaleTimesheetController(http.Controller):
...
+        dashboard_values['profit'] = profit
 
         values['dashboard'] = dashboard_values
 
-        # -- Time Repartition (per employee)
-        repartition_domain = domain + [('employee_id', '!=', False), ('timesheet_invoice_type', '!=', False)]  # force billable type
+        #
+        # Time Repartition (per employee per billable types)
+        #
+        employees = projects.mapped('tasks.user_id.employee_ids') | request.env['account.analytic.line'].search([('project_id', 'in', projects.ids)]).mapped('employee_id')
```
Comment (tde-banana-odoo) [inline addons/sale_timesheet/controllers/main.py]: [Commit3] Small perf proposal here: is the compute of employees based on aal really necessary ? Here below in repartition_employee you do a setdefault. I am therefore not sure it is necessary to compute the whole employees. Just having those coming from tasks in order to have them listed even if they made no timesheet seems sufficient.

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 2]
```diff
@@ -64,6 +65,8 @@ def _timesheet_preprocess(self, vals):
         if vals.get('project_id') and not vals.get('account_id'):
             project = self.env['project.project'].browse(vals.get('project_id'))
             vals['account_id'] = project.analytic_account_id.id
+            if not project.analytic_account_id.active:
```
Comment (tde-banana-odoo) [inline addons/hr_timesheet/models/hr_timesheet.py]: [Commit4] Small question while we are at it: is there any logic and/or propagation implemented about archiving stuff ? Like archiving a project, unarchiving a project, archiving analytic account, ... ?

EDIT: I see there are stuff displayed on project and task views, so I suppose it is ok and mighty bst and you took it into consideration :) .

File: addons/hr_timesheet/models/project.py [hunk 3]
```diff
@@ -1,31 +1,99 @@
...
+        values = {
+            'name': name,
+            'allow_timesheets': True,
+        }
+        return self.create(values).name_get()[0]
+
+    @api.model
+    def create(self, values):
+        """ Create an analytic account if project allow timesheet and don't provide one """
+        if values.get('allow_timesheets') and not values.get('analytic_account_id'):
```
Comment (tde-banana-odoo) [inline addons/hr_timesheet/models/project.py]: [Commit4] Don't forget to include my magnificient fix. Just throwing some flowers at myself.

File: addons/hr_timesheet/models/project.py [hunk 4]
```diff
@@ -1,31 +1,99 @@
...
+        for project in self:
+            if project.analytic_account_id and not project.analytic_account_id.line_ids:
+                analytic_accounts_to_delete |= project.analytic_account_id
+        result = super(Project, self).unlink()
+        analytic_accounts_to_delete.unlink()
+        return result
+
+    def _create_analytic_account(self, values):
+        analytic_account = self.env['account.analytic.account'].create({
+            'name': values.get('name', _('Unkwon Analytic Account')),
```
Comment (tde-banana-odoo) [inline addons/hr_timesheet/models/project.py]: [Commit4] Note sure: nothing about currency ? I suppose it comes from company and therefore not necessary to re-specify it ?

File: addons/sale_timesheet/controllers/main.py [hunk 5]
```diff
@@ -97,23 +116,171 @@ def _prepare_plan_values(self, domain):
...
+
+        def _to_short_month_name(date):
+            month_index = fields.Date.from_string(date).month
+            return babel.dates.get_month_names('abbreviated', locale=request.env.context.get('lang', 'en_US'))[month_index]
+
+        header = [_('Name'), _('Before')] + [_to_short_month_name(date) for date in ts_months] + [_('Done'), _('Sold'), _('Remaining')]
+        return header
+
+    def _table_row_default(self, projects):
+        return [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]  # before, M1, M2, M3, Done, Sold, Remaining
```
Comment (tde-banana-odoo) [inline addons/sale_timesheet/controllers/main.py]: To try to be a bit less hardcoded, maybe you could simply return [0.0] * len(table_header - 1)

File: addons/sale_timesheet/controllers/main.py [hunk 6]
```diff
@@ -12,74 +12,87 @@ class SaleTimesheetController(http.Controller):
 
     @http.route('/timesheet/plan', type='json', auth="user")
     def plan(self, domain):
-        domain = expression.AND([domain, [('project_id', '!=', False)]])  # force timesheet and not AAL
-        values = self._prepare_plan_values(domain)
+        """ Get the HTML of the project plan for projects matching the given domain
+            :param domain: a domain for project.project
+        """
+        projects = request.env['project.project'].search(domain)
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/controllers/main.py]: Passing the domain will mean that each override of `_prepare_plan_values` will execute a `search(domain)` to get the concerned projects and extract informations from them. So multiple search with different prefetch (to be confirm).
What do you think ?

File: addons/sale_timesheet/controllers/main.py [hunk 7]
```diff
@@ -12,74 +12,87 @@ class SaleTimesheetController(http.Controller):
...
+        values = self._prepare_plan_values(projects)
         view = request.env.ref('sale_timesheet.timesheet_plan')
         return {
             'html_content': view.render(values)
         }
 
-    def _prepare_plan_values(self, domain):
+    def _prepare_plan_values(self, projects):
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/controllers/main.py]: I can rename `_prepare_plan_values` into `_plan_prepare_value`  to be consistent with `_plan_get_stat_button` ;)

File: addons/sale_timesheet/controllers/main.py (hunk 1 above)
Comment (jem-odoo) [reply addons/sale_timesheet/controllers/main.py]: Unfortunately for performance, the Overview must contain all employee assigned to task, and all employee having timesheeted in the project ...
In the spec, it says "Time by people : Display all the users that have been assigned to tasks (time = 0 if no timesheet) TO CHECK : DOES IT ALSO DISPLAY THE EMPLOYEES WHO DID SOME TIMESHEETS ON THE TASKS ????  yes it does, nice job "

File: addons/hr_timesheet/models/project.py (hunk 4 above)
Comment (jem-odoo) [reply addons/hr_timesheet/models/project.py]: currency field is a related on the company.

File: addons/sale_timesheet/wizard/project_task_assign_so_line.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class ProjectTaskAssignSOLine(models.TransientModel):
+    _name = 'project.task.assign.sale.order.line'
+    _description = "Assign Sale Order line to tasks"
+
+    @api.model
+    def default_get(self, fields):
+        result = super(ProjectTaskAssignSOLine, self).default_get(fields)
+        active_ids = self._context.get('active_ids')
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/wizard/project_task_assign_so_line.py]: Should we raise an exception if active_mode is not task ?

File: addons/sale_timesheet/wizard/project_task_assign_so_line.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class ProjectTaskAssignSOLine(models.TransientModel):
+    _name = 'project.task.assign.sale.order.line'
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/wizard/project_task_assign_so_line.py]: I guess you don't want to see the 'wizard' word in the name, so what do you have in mind ?

File: addons/sale_timesheet/controllers/main.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/sale_timesheet/controllers/main.py]: NICE JOB !

File: addons/hr_timesheet/models/project.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/hr_timesheet/models/project.py]: Ok then.

File: addons/sale_timesheet/wizard/project_task_assign_so_line.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sale_timesheet/wizard/project_task_assign_so_line.py]: I would just skip the active_ids part if active_model is incorrect.

File: addons/sale_timesheet/controllers/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sale_timesheet/controllers/main.py]: Ok then. Let us avoid prefetch hell.

File: addons/sale_timesheet/wizard/project_task_assign_so_line.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sale_timesheet/wizard/project_task_assign_so_line.py]: project.task.assign.sale ?

File: addons/sale_timesheet/controllers/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sale_timesheet/controllers/main.py]: Dev will be less sad.

File: addons/sale_timesheet/wizard/project_task_assign_so_line.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/wizard/project_task_assign_so_line.py]: ok, will do ;)

**Comment by bst-odoo** [comment]
WOUHOU !

**Comment by jem-odoo** [comment]
IMHO, Migration should not be so hard, since only 4 commits impact database schema. Here are some tips:

* [IMP] hr_timesheet: ability to track cost without selling time
timesheet_cost and currency_id are moved from sale_timesheet to hr_timesheet (always on hr.employee)
* [IMP] project: a project is not an analytic account anymore !
inhertiS between project.project and account.analytic.account is removed
module dependencies: hr_timesheet now depends on analytic, and project does not depends on analytic anymore
fields name, partner_id, company_id, currency_id should be duplicated from the analytic account to the project (as those columns are now existing on the project table)
Note: an analytic account can now have multiple project. Maybe some customised database already have that, but now it is completely supported.
For existing projects, keep the AA set (even if they not allow_timesheet) to still be able to reinvoice expenses on the project (only when hr_timesheet is installed)
* [REF] sale_timesheet: project/task generation from SO 
task_id already exists on sale.order.line and contains the initial generated task by the SOL. The same should apply for the new project_id field: we can populate it (if the SOL product should create a project aka service_tracking in 'task_new_project' or 'project_only') with the first project (create_date ASC) having sale_line_id filled with the current SOL. This is like a one2one between project.project and sale.order.line.
* [IMP] sale_timesheet: group task by Sales Order
related stored field, should be handled by the ORM

**Comment by nhomar** [comment]
Hello @jem-odoo 

My dear friend, falling apart of the nice set of improvements for new databases this is bringing (I am not against any change) I would like to point you out the **migration** is not only a matter of "Data Structure" it is about, data usage, meaning of the data and functional management of the change.

With this changes, the good news is that we will be able to use project **only** which is a great step forward.

The things I am worried about are:

1. Before a project was a "Team" not a project per sé, the inheritance help to have that linked, now that it is different, "What will be the approach taken for those which the project was a team and not a project per sé?" >> Just a short answer is enough, and that answer should be linked to the migration script.
2. The relation between task and sale.order.line is not one2one always, you can set a ton of tasks to solve one quoted sale order line (functionally, conceptually and in practice), if I understood well (maybe I am wrong) it is not aligned with the new approach (correct me please if I am wrong).
3. For me (since so much time ago) a project is a cost center, this  **Note: an analytic account can now have multiple projects. Maybe some customized database already have that, but now it is completely supported.** can be complex to migrate to, what if a DB has this customization? what if this was solved with another functional approach? for example in our case we simply allow set the AA into the timesheet (even if different from the task itself) with that we sot such **issue**, you need to consider that, (no customization is done more than in views).

I will be thinking in other topics, but the point is that you should start to think on the migration path to help us to have it in 1 year from now ready to go.

Remember now we are a 10 years system, we have more than 10k tasks, already organized to report with, and so on, this kind of change will impact me and others I think just write down in the interim the situations to be solved (I will do it) and I know we will figure it out.

Regards and happy hacking.

**Pull request merged**